vllm_xpu_kernels/fused_moe_interface.py

```diff
@@ -0,0 +1,183 @@
+                      n_experts_per_token, activation, num_experts):
+
+    token_cnt, hidden_size = list(hidden_states.shape)
+    intermediate_size = list(w2.shape)[-1]
+    total_input_size = token_cnt * n_experts_per_token
+    if not hasattr(cutlass_fused_moe, "moe_buffer"):
+        moe_buffer = {}
+        device = hidden_states.device
+        moe_buffer["expert_cache"] = torch.empty((token_cnt * hidden_size),
+                                                 dtype=hidden_states.dtype,
+                                                 device=hidden_states.device)
+        moe_buffer["gemm1_input"] = torch.empty(
+            (total_input_size, hidden_size),
+            dtype=hidden_states.dtype,
+            device=hidden_states.device)
+        moe_buffer["gemm1_output"] = torch.empty(
+            (total_input_size, 2 * intermediate_size),
+            dtype=hidden_states.dtype,
+            device=hidden_states.device)
+        moe_buffer["gemm2_output"] = torch.empty(
+            (total_input_size, hidden_size),
+            dtype=hidden_states.dtype,
+            device=hidden_states.device)
+        moe_buffer["alpha"] = torch.ones(num_experts,
+                                         dtype=torch.float32,
+                                         device=hidden_states.device)
+        moe_buffer["beta"] = torch.zeros(num_experts,
+                                         dtype=torch.float32,
+                                         device=hidden_states.device)
+
+        cutlass_fused_moe.moe_buffer = moe_buffer
+
+    expert_cache = cutlass_fused_moe.moe_buffer[
+        "expert_cache"][:hidden_states.numel()].view_as(hidden_states).zero_()
+    input_A = cutlass_fused_moe.moe_buffer["gemm1_input"][:total_input_size, :]
+    gemm1_output = cutlass_fused_moe.moe_buffer[
+        "gemm1_output"][:total_input_size, :]
+    gemm2_output = cutlass_fused_moe.moe_buffer[
+        "gemm2_output"][:total_input_size, :]
+    alpha = cutlass_fused_moe.moe_buffer["alpha"]
+    beta = cutlass_fused_moe.moe_buffer["beta"]
+
+    # map token to experts
+    idxs = topk_ids.argsort()
+    counts = topk_ids.to(torch.long).bincount().cpu().numpy()
+    tokens_per_expert = counts.cumsum()
+    num_per_tok = n_experts_per_token
+    token_idxs = idxs // num_per_tok
+    offset = []
+    for expert_id, end_idx in enumerate(tokens_per_expert):
+        start_idx = 0 if expert_id == 0 else tokens_per_expert[expert_id - 1]
+        offset.append(end_idx - start_idx)
+        if start_idx == end_idx:
+            continue
+        exp_token_idxs = token_idxs[start_idx:end_idx]
+        # expert_tokens = hidden_states[exp_token_idxs]
+        # grouped_input_A.append(expert_tokens)
+        input_A[start_idx:end_idx, :].copy_(hidden_states[exp_token_idxs])
+
+    while len(offset) < num_experts:
+        offset.append(0)
+
+    ########### gemm1 ##################
+    input_B = w13.transpose(-1, -2).contiguous().transpose(-1, -2)
```

Comment:
You can use `ColumnMajor` `B` in the GroupGEMM kernel, so that you wouldn't have to transpose `B`.
To use column-major `B` (to avoid transposing weights), you can use a different copy atom for transposed loads . Coincidentally, because of how thread-value assignment works in the copy-atoms, the transpose copy atom for 16-bit (or 8-bits, for that matter) dtypes will load data in VNNI format (which is also true for atoms ending in `_N`).

However, if the latency of `tranpose of B` + `GEMM with RowMajor B` is lower than `GEMM with columnMajor B`, then you might want to retain this approach.

FWIW, once 32x32 transpose copy-atom for BF16 is added in cutlass, perf of `GEMM with columnMajor B` will become a bit better.

Thanks!